zzre/game/systems/ui/ScrBookMenu.cs

```diff
@@ -42,27 +72,111 @@ protected override void HandleOpen(in messages.ui.OpenBookMenu message)
             .Build();
 
         CreateTopButtons(preload, entity, inventory, IDOpenFairybook);
-        CreateFairyButtons(preload, entity, inventory);
+        CreateFairyButtons(preload, entity, inventory, ref book);
     }
 
-    private void CreateFairyButtons(UIPreloader preload, in DefaultEcs.Entity entity, Inventory inventory)
+    private void CreateFairyButtons(UIPreloader preload, in DefaultEcs.Entity entity, Inventory inventory, ref components.ui.ScrBookMenu book)
     {
-        var fairies = db.Fairies.OrderBy(fairyRow => fairyRow.CardId.EntityId).ToArray();
+        var fairies = book.Fairies;
         for (int i = 0; i < fairies.Length; i++)
         {
             if (inventory.Contains(fairies[i].CardId))
             {
                 var element = new components.ui.ElementId(1 + i);
-                preload.CreateButton(entity)
+                var button = preload.CreateButton(entity)
                     .With(element)
-                    .With(Mid + new Vector2(226 + 45 * (i % 9), 66 + 45 * (i / 9)))
+                    .With(Mid + FairyButtonPos(i))
                     .With(new components.ui.ButtonTiles(fairies[i].CardId.EntityId))
                     .With(preload.Wiz000)
                     .Build();
+                book.FairyButtons.Add(element, fairies[i]);
+
+                // If player has multiple fairies of the same card id, only the first caught counts
+                if (inventory.Fairies.First(c => fairies[i].CardId == c.cardId).isInUse) {
+                    preload.CreateImage(entity)
+                        .With(Mid + FairyButtonPos(i))
+                        .With(preload.Inf000, 16)
+                        .WithRenderOrder(-1)
+                        .Build();
+                }
             }
         }
     }
 
+    private DefaultEcs.Entity CreateSidebar(UIPreloader preload, in DefaultEcs.Entity parent, FairyRow fairyRow, ref components.ui.ScrBookMenu book)
+    {
+        var entity = World.CreateEntity();
+        entity.Set(new components.Parent(parent));
+
+        var fairyI = Array.FindIndex(book.Fairies, row => row.Name == fairyRow.Name) + 1;
```

Comment:
This will work but as `FairyRow` is a reference type you can just use `Array.IndexOf`